examples/gno.land/r/sys/validators/validators.gno

```diff
@@ -45,33 +49,63 @@ func removeValidator(address std.Address) {
 	}
 
 	// Validator removed, note the change
-	changes = append(changes, change{
+	ch := change{
 		blockNum: std.GetHeight(),
 		validator: validators.Validator{
 			Address:     val.Address,
 			PubKey:      val.PubKey,
 			VotingPower: 0, // nullified the voting power indicates removal
 		},
-	})
+	}
+
+	saveChange(ch)
 
 	// Emit the validator set change
 	std.Emit(validators.ValidatorRemovedEvent)
 }
 
+// saveChange saves the valset change
+func saveChange(ch change) {
+	id := getBlockID(ch.blockNum)
+
+	setRaw, exists := changes.Get(id)
+	if !exists {
+		setRaw = make([]change, 0, 1)
+	}
+
+	// Save the change
+	set := setRaw.([]change)
+	set = append(set, ch)
+
+	changes.Set(id, set)
+}
+
+// getBlockID converts the block number to a sequential ID
+func getBlockID(blockNum int64) string {
+	return seqid.ID(uint64(blockNum)).String()
+}
+
 func Render(_ string) string {
-	if len(changes) == 0 {
+	if changes.Size() == 0 {
 		return "No valset changes to apply."
 	}
 
 	output := "Valset changes:\n"
-	for _, change := range changes {
-		output += ufmt.Sprintf(
-			"- #%d: %s (%d)\n",
-			change.blockNum,
-			string(change.validator.Address),
-			change.validator.VotingPower,
-		)
-	}
+
+	changes.Iterate("", "", func(_ string, value interface{}) bool {
```

Comment:
```go
size := changes.Size()
changes.ReverseIterateByOffset(size, size-10, ...)
```

to get the last 10 entries.

Reply:
Do you think we should render just the last 10 entries, and not all of them?

Or should we use the `Render` param as the "from"?

Reply:
The last N entries are ideal, in my opinion. I dislike the current approach of never including pagination or limits in contracts, as we'll end up with numerous contracts that will become unwieldy over time. Additionally, people who copy-paste our contracts to write their own will face the same problems. We should show good practices.

Purely in terms of usage for this contract, the last N entries are sufficient, in my opinion. We can then use indexers to access the longer history.

Once we have a good pagination library, you can simply plug it in to enable automatic pagination. For now, please keep things simple and return the last N entries.

Reply:
Updated to show the last 10 entries:

abc2fc1